subsys/net/lib/app/net_app.c

```diff
@@ -157,17 +157,27 @@ void _net_app_received(struct net_context *net_ctx,
 #if defined(CONFIG_NET_APP_SERVER)
 	if (ctx->app_type == NET_APP_SERVER) {
 		if (!pkt) {
-			if (ctx->cb.close) {
-				ctx->cb.close(ctx, status, ctx->user_data);
-			}
-
 #if defined(CONFIG_NET_TCP)
-			if (ctx->proto == IPPROTO_TCP) {
-				net_context_put(ctx->server.net_ctx);
-				ctx->server.net_ctx = NULL;
+			int i;
+
+			for (i = 0;
+			     ctx->proto == IPPROTO_TCP &&
+				     i < CONFIG_NET_APP_SERVER_NUM_CONN;
+			     i++) {
+				if (ctx->server.net_ctxs[i] == net_ctx &&
+				    ctx == net_ctx->net_app) {
+					net_context_put(net_ctx);
+					ctx->server.net_ctxs[i] = NULL;
+					net_ctx->net_app = NULL;
+					break;
+				}
 			}
 #endif
 
+			if (ctx->cb.close) {
```

Comment:
Is it ok that we're reversing the order of the close callback here?  (I'll reference this later in the review)